File: GA.py
from random import random,randint,shuffle,choice,sample,randrange
from sys import setrecursionlimit,stdout
from time import sleep, time
from os import system
setrecursionlimit = 100000


write = stdout.write
######################### Funcoes de validacao #########################
def isValid(x,y):
	if 0 <= x < mazeSize and 0 <= y < mazeSize:
		return True
	return False

def isAWall(x,y):
	return maze[x][y] == wallIcon 

########################################################################

################### Informacoes/Funcoes do labirinto ###################
# Algoritmo de Gilmar totalmente corrigido pelo proprio, gera um labirinto chamado maze de tamanho sizeXsize
# uma matriz chamada mazeDists de tamanho sizeXsize que possui as distancias de cada ponto para a saida.
def generateMaze(mazeSize):
	x,y = randrange(1,mazeSize, 2),(mazeSize-1)*randint(0, 1)
	coords = [(x,x),(y, y-1 if y else 1)]
	shuffle(coords)
	dfsStart, exit, spawn = (coords[0][1], coords[1][1]), (coords[0][0],coords[1][0]), (randrange(1, mazeSize, 2), randrange(1, mazeSize, 2))
	maze[dfsStart[0]][dfsStart[1]] = ' '
	maze[exit[0]][exit[1]] = 'E'
	mazeDists[exit[0]][exit[1]] = 0
	generateMazeRecursive(dfsStart[0],dfsStart[1],1)
	maze[spawn[0]][spawn[1]] = 'S'
	return spawn,exit

def generateMazeRecursive(y,x,dist):
	mazeDists[y][x] = dist
	pairs = [(1,0),(0,1),(-1,0),(0,-1)]
	shuffle(pairs)
	for i in xrange(4):
		if isValid(y+2*pairs[i][0],x+2*pairs[i][1]) and isAWall(y+2*pairs[i][0],x+2*pairs[i][1]):
			maze[y+pairs[i][0]][x+pairs[i][1]] = maze[y+2*pairs[i][0]][x+2*pairs[i][1]] = ' '
			mazeDists[y+pairs[i][0]][x+pairs[i][1]] = dist+1
			generateMazeRecursive(y+2*pairs[i][0],x+2*pairs[i][1], dist+2)

def drawMaze(maze):
	slp = 0
	for i in xrange(mazeSize):
		separator = " "
		if not i or i == mazeSize-1:
			separator = wallIcon
		print separator.join(maze[i])
		sleep(slp)

########################################################################

################### Informacoes/Funcoes de movimento ###################
def vectorSum(point,direction):
	return (point[0]+direction[0],point[1]+direction[1])

def makeAMove(point,direction):
	return vectorSum(point,moves[direction])

def isValidMove(point,direction):
	newPoint = vectorSum(point,moves[direction])
	return isValid(newPoint[0],newPoint[1]) and not (isAWall(newPoint[0],newPoint[1]))


directions = ['U','D','L','R']
moves = {'U':(-1,0),'D':(1,0),'L':(0,-1),'R':(0,1)}
coherentMoves = {'U':['U','L','R'],'D':['D','L','R'],'R':['R','D','U'],'L':['L','D','U']}

######################  Informacoes/Funcoes do GA ######################

def calculateFitness():
	""" A funcao fitness eh calculada a partir de 3 principios:
	1 - Cada passo dado recebe uma penalizacao de 0.05 pontos (Retirado para testar se estava influenciando negativamente)
	2 - Cada choque com a parede recebe um penalizacao de 2.0 pontos
	3 - Cada vez que um individuo volta a uma posicao ja visitada ele recebe uma penalizacao de 1.0 pontos
	Individuos que nao se movem ,ou seja, que tem fitness igual a 0 tem o fitness reduzido drasticamente .
	Individuos que acham a saida tem o fitness aumentado drasticamente.
	"""
	for k in xrange(populationSize):
		individual = population[k][1]
		startPoint = spawn
		fitness = fitnessConstant
		visited = set([spawn])

		for move in individual:
			if move is not None:
				if isValidMove(startPoint,move):
					startPoint = makeAMove(startPoint,move)
					fitness += 10
				else:
					fitness -= 200

				if startPoint in visited:
					fitness -= 300
				else:
					visited.add(startPoint)

				if startPoint == exit:
					fitness *= fitnessConstant
					break

		if not individual:
			fitness = -fitnessConstant
		
		## Mudei
		fitness -= mazeDists[startPoint[0]][startPoint[1]]*1000

		individualsFitness[k] = fitness

def getFitness(index):
	return individualsFitness[index]
 
def sortByFitness():
	population.sort(key = lambda pop : (individualsFitness[pop[0]],-len(pop[1])), reverse = True)

def getFittest():
	sortByFitness()
	return population[0][1]

## Mudei
def initPopulation():
	for index in xrange(populationSize):
		counter = 0
		for move in xrange(randint(1,chromossomeSize)):
			if move:population[index][1][counter] = choice(coherentMoves[population[index][1][counter-1]])
			else: population[index][1][counter] = choice(directions)
			counter += 1

def crossOver():
	newPopulation = []
	cromossomesSet = set()

	for i in xrange(populationSize):
		# Uma forma alternativa de contornar os casos de convergencia prematura eh impedir que solucoes iguais coexistam
		# soh que garantir isso eh bem custoso.
		sonIsValid = False
		while not sonIsValid:
			son = []
			l1,r1 = groups[choice(groupsArray)]
			l2,r2 = groups[choice(groupsArray)]

			daddy = population[randint(l1,r1)][1]
			mommy = population[randint(l2,r2)][1]

			if not l1 and l2:
				daddy = filter(lambda x: x is not None, daddy)
				for _ in xrange(chromossomeSize/2):
					if not _: daddy.append(choice(directions))
					else: daddy.append(choice(coherentMoves[daddy[-1]]))
				son = daddy

			elif not l2 and l1:
				mommy = filter(lambda x: x is not None, mommy)
				for _ in xrange(chromossomeSize/2):
					if not _: mommy.append(choice(directions))
					else: mommy.append(choice(coherentMoves[mommy[-1]]))
				son = mommy

			else:
				first,second = sample([daddy,mommy],2)
				crossOverPoint = randint(1,chromossomeSize) 
				
				for k in xrange(crossOverPoint):
					if k < len(first):
						if first[k] is None: 
							crossOverPoint = k
							break
						son.append(first[k])
						

				for j in xrange(crossOverPoint,chromossomeSize):
					if j < len(second):
						if second[j] is None: break
						#Mudei
						if j == crossOverPoint: son.append(choice(coherentMoves[son[-1]]))
						else: son.append(second[j])

			sonString = ''.join(son)
			if sonString not in cromossomesSet:
				sonIsValid = True
				cromossomesSet.add(sonString)

		mutation(son)
		newPopulation.append((i,son))


	return newPopulation

## Mudei
def mutation(cromossome):
	if mutationChance > random() and cromossome:
		if random() > 0.6:
			for i in xrange(randint(1,chromossomeSize/2)):
				if cromossome: cromossome.append(choice(coherentMoves[cromossome[-1]]))
				else : cromossome.append(choice(directions))
		else:
			for i in xrange(randint(1,chromossomeSize)):
				idx = randint(0,len(cromossome)-1)
				if idx: cromossome[idx] = choice(coherentMoves[cromossome[idx-1]])
				else: cromossome[idx] = choice(directions)

def drawFittestMoves(fittest,generation):
	drawMaze(maze)
	mazeCopy = maze
	move = spawn
	for direction in fittest:
		if direction is not None and isValidMove(move,direction): 
			mazeCopy[move[0]][move[1]] = ' '
			move = makeAMove(move,direction)
			if move == exit: break
			mazeCopy[move[0]][move[1]] = 'M'
			system('cls||clear')
			drawMaze(mazeCopy)
			print "Generation %i Fittest Fitness = %.0f" %(generation,fittestFitness%fitnessConstant)
			sleep(0.3)
			
	sleep(1)

########################################################################
print('''
----------------------------------------------------------
#################### - MAZER SOLVER - ####################
----------------------------------------------------------
''')

mazeSize = int(raw_input('Size of maze?') or '15')

#mazeSize = 15 # line/column
wallIcon = '#'

while(True):
	maze = [[wallIcon]*mazeSize for i in xrange(mazeSize)]
	mazeDists = [[-1]*mazeSize for i in xrange(mazeSize)]

	spawn,exit = generateMaze(mazeSize)
	drawMaze(maze)

	genNew = raw_input("Generate new maze?(y/n) ")
	if(genNew.strip() == 'n'):
		break


numOfWalls = sum([maze[i].count(wallIcon) for i in xrange(mazeSize)])

populationSize = int(raw_input('Size of population: ') or '1000')

#populationSize = 1000
chromossomeSize = mazeSize ** 2 - numOfWalls -1
population = [(i,[None]*chromossomeSize) for i in xrange(populationSize)]
individualsFitness = [-1]*populationSize
mutationChance = 0.4
fitnessConstant = 10e6

# Grupos do fitness
quartil = [populationSize*0.01,populationSize*0.25,populationSize*0.4,populationSize*0.75]
groups = {1:(0,quartil[0]),
		  2:(quartil[0]+1,quartil[1]),
		  3:(quartil[1]+1,quartil[2]),
		  4:(quartil[2]+1,quartil[3]),
		  5:(quartil[3]+1,populationSize-1)}
groupsChance = [0.5,0.25,0.15,0.08,0.02]
groupsArray = [1]*(int(groupsChance[0]*100)) + [2]*(int(groupsChance[1]*100)) + [3]*(int(groupsChance[2]*100)) + [4]*(int(groupsChance[3]*100))

initPopulation()
fittest = generation = fittestFitness = 0
count = 0
finalists = []


startTime = time()

# Esse count eh soh para efeitos de teste
while count < 20000 and len(finalists) < 5:
	calculateFitness()
	fittest = getFittest()
	fittestFitness = getFitness(population[0][0])

	if fittestFitness > fitnessConstant:
		finalists.append(fittest)
	print len(finalists)

	population = crossOver()
	
	generation += 1
	count +=1
	
	print 'Solving','.'*(generation%5)
	sleep(0.5)
	system('cls||clear')

endTime = time()

exec_time = endTime-startTime
finalists.sort(key = lambda x: len(x))

fittest = finalists[0]

sleep(1)
drawFittestMoves(fittest,generation)
print("Time Executation: %.2f seconds" %(exec_time))
input()

File: GA.cpp
#include <bits/stdc++.h>
#include <windows.h>
#define put push_back
#define F first
#define S second
using namespace std;
typedef pair<pair<int, int>, pair<int, int>> pairOfPairs;
typedef struct individual;

/*
///////////////////////////////////////////////
Variaveis iniciais
*/
int mazeSize = 0;
int populationSize = 1000;
const double mutationChance = 0.4;
const int fitnessConstant = 10e6;
vector<individual> population;
char **maze;
int **mazeDists;
char wallIcon = '|';
pair<int, int> spawn, escape;
int numOfWalls = 0;
char directions[4] = {'U', 'D', 'L', 'R'};
map<char, pair<int, int>> moves = {{'U', {-1, 0}}, {'D', {1, 0}}, {'L', {0, -1}}, {'R', {0, 1}}};
int chromossomeSize;
map<char, array<char, 3>> coherentMoves = {{'U', {'U', 'L', 'R'}}, {'D', {'D', 'L', 'R'}}, {'R', {'R', 'D', 'U'}}, {'L', {'L', 'D', 'U'}}};
int quartil[4] = {(int)(populationSize * 0.01), (int)(populationSize * 0.25), (int)(populationSize * 0.4), (int)(populationSize * 0.75)};
map<int, pair<int, int>> groups = {{1, {0, quartil[0]}}, {2, {quartil[0] + 1, quartil[1]}}, {3, {quartil[1] + 1, quartil[2]}}, {4, {quartil[2] + 1, quartil[3]}}, {5, {quartil[3] + 1, populationSize - 1}}};
double groupsChance[5] = {0.5, 0.25, 0.15, 0.08, 0.02};
int groupsArray[100];

/*
///////////////////////////////////////////////
Structs iniciais
*/

struct individual
{
    int fitness;
    vector<char> moves;
    individual() { fitness = 0; }
    bool operator<(individual other)
    {
        //Comparador utilizado no sort
        return fitness < other.fitness;
    }
};

/*
///////////////////////////////////////////////
Funcoes auxiliares
*/

bool isValid(int x, int y)
{
    bool xValid = (0 <= x && x < mazeSize);
    bool yValid = (0 <= y && y < mazeSize);
    return xValid && yValid;
}

bool isAWall(int x, int y)
{
    return maze[x][y] == wallIcon;
}

void randomSeed()
{
    srand((int)time(0));
}

int randomRange(int start, int end, int step = 1)
{
    if (start > end){
        int temp = start;
        start = end;
        end = start;
    }
    return (((rand() % (end - start + 1)) / step) * step) + start;
}

void initVariables()
{
    maze = new char *[mazeSize];
    mazeDists = new int *[mazeSize];
    for (int i = 0; i < mazeSize; i++)
    {
        maze[i] = (char *)malloc(mazeSize * sizeof(char));
        memset(maze[i], wallIcon, mazeSize * sizeof(char));
        mazeDists[i] = (int *)malloc(mazeSize * sizeof(int));
        memset(mazeDists[i], -1, mazeSize * sizeof(int));
    }
    // double groupsChance[5] = {0.5, 0.25, 0.15, 0.08, 0.02};
    int j = 5;
    int count = 0;
    for (int i = 99; i >= 0; i--)
    {
        if (count < groupsChance[j-1]*100 ){
            count++;
            groupsArray[i] = j;
        }
        else {
            count = 0;
            j--;
            i++;
        }
    }
}

/*
///////////////////////////////////////////////
Funcoes utilitarias
*/

void mazeGeneratorRecursive(int x, int y, int dist)
{
    mazeDists[x][y] = dist;
    int pairs[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    random_shuffle(pairs, pairs + 4);
    for (int i = 0; i < 4; i++)
    {
        if (isValid(x + 2 * pairs[i][0], y + 2 * pairs[i][1]) && isAWall(x + 2 * pairs[i][0], y + 2 * pairs[i][1]))
        {
            maze[x + pairs[i][0]][y + pairs[i][1]] = ' ';
            maze[x + 2 * pairs[i][0]][y + 2 * pairs[i][1]] = ' ';
            mazeDists[x + pairs[i][0]][y + pairs[i][1]] = dist + 1;
            numOfWalls -= 2;
            mazeGeneratorRecursive(x + 2 * pairs[i][0], y + 2 * pairs[i][1], dist + 2);
        }
    }
}
void mazeGenerator()
{
    
    int x = randomRange(1, mazeSize - 1, 2), y = (mazeSize - 1) * randomRange(0, 1);
    pairOfPairs coords = {{x, y}, {x, (y) ? y - 1 : 1}};
    if (randomRange(0, 1))
    {
        swap(coords.F.F, coords.F.S);
        swap(coords.S.F, coords.S.S);
    }
    escape = coords.F, spawn = {randomRange(1, mazeSize - 1, 2), randomRange(1, mazeSize - 1, 2)};
    maze[escape.F][escape.S] = 'E', maze[coords.S.F][coords.S.S] = ' ';
    mazeDists[escape.F][escape.S] = 0;
    numOfWalls -= 2;
    mazeGeneratorRecursive(coords.S.F, coords.S.S, 1);
    maze[spawn.F][spawn.S] = 'S';
    chromossomeSize = mazeSize * mazeSize - numOfWalls - 1;
}

void drawMaze()
{
    for (int i = 0; i < mazeSize; i++)
    {
        for (int j = 0; j < mazeSize; j++)
        {
            putchar(maze[i][j]);
        }
        putchar('\n');
    }
}

void initPopulation()
{
    for (int ind = 0; ind < populationSize; ind++)
    {
        population.push_back(individual());
        for (int move = 0; move < (rand() % chromossomeSize + 1); move++)
        {
            if (move)
            {
                population[ind].moves.push_back(coherentMoves[population[ind].moves.back()][randomRange(0, 2)]);
            }
            else
            {
                population[ind].moves.push_back(directions[randomRange(0, 3)]);
            }
        }
    }
}

/*
///////////////////////////////////////////////
Metodos de movimento
*/

pair<int, int> vectorSum(pair<int, int> point, pair<int, int> direction)
{
    pair<int, int> newPoint = {point.F + direction.F, point.S + direction.S};
    return newPoint;
}

pair<int, int> makeAMove(pair<int, int> point, char direction)
{
    return vectorSum(point, moves[direction]);
}

bool isValidMove(pair<int, int> point, char direction)
{
    pair<int, int> newPoint = vectorSum(point, moves[direction]);
    return isValid(newPoint.F, newPoint.S) && isAWall(newPoint.F, newPoint.S);
}

/*
///////////////////////////////////////////////
Metodos do GA
*/

void calculateFitness()
{
    for (int i = 0; i < populationSize; i++)
    {
        individual indiv = population[i];
        set<pair<int, int>> visited;
        long long fitness = fitnessConstant;
        pair<int, int> startPoint = spawn;
        visited.insert(spawn);
        bool finished = false;

        // Talvez seja necessario colocar uma verificacao de not null aqui
        for (int j = 0; j < indiv.moves.size(); j++)
        {
            char move = indiv.moves[j];

            if (isValidMove(startPoint, move))
            {
                startPoint = makeAMove(startPoint, move);
                fitness += 10;
            }
            else
            {
                fitness -= 200;
            }

            if (!visited.insert(startPoint).S)
            {
                fitness -= 300;
            }

            if (startPoint == escape)
            {
                fitness *= fitnessConstant;
                finished = true;
                break;
            }
        }

        if (indiv.moves.empty())
        {
            fitness = -fitnessConstant;
        }

        fitness -= mazeDists[startPoint.F][startPoint.S] * 1000;
        indiv.fitness = fitness;
    }
}

void sortByFitness()
{
    reverse(population.begin(), population.end());
}

vector<char> getFittest()
{
    sortByFitness();
    return population.front().moves;
}

double doubleRand()
{
    return double(rand()) / (double(RAND_MAX) + 1.0);
}

void mutation(individual &cromossome)
{
    if (mutationChance > doubleRand() && !cromossome.moves.empty())
    {
        if (doubleRand() > 0.6)
        {
            int temp = randomRange(1, cromossome.moves.size() / 2);
            for (int i = 0; i < temp; i++)
            {
                if (!cromossome.moves.empty())
                {
                    // coherentMoves[mommy.back()][randomRange(0, 2)]
                    cromossome.moves.push_back(coherentMoves[cromossome.moves.back()][randomRange(0, 2)]);
                }
                else
                {
                    cromossome.moves.push_back(directions[randomRange(0, 3)]);
                }
            }
        }
        else
        {
            int temp = randomRange(1, cromossome.moves.size());
            for (int i = 0; i < temp; i++)
            {
                int idx = randomRange(0, cromossome.moves.size() - 1);
                if (idx != 0)
                {
                    cromossome.moves[idx] = coherentMoves[cromossome.moves[idx - 1]][randomRange(0, 2)];
                }
                else
                {
                    cromossome.moves[idx] = directions[randomRange(0, 3)];
                }
            }
        }
    }
}

vector<individual> crossover()
{
    vector<individual> newPopulation;
    set<string> chromossomeSet;

    for (int i = 0; i < populationSize; i++)
    {
        bool sonIsValid = false;
        vector<char> son;
        while (!sonIsValid)
        {
            pair<int, int> pairL = groups[groupsArray[randomRange(0, 99)]];
            pair<int, int> pairR = groups[groupsArray[randomRange(0, 99)]];
            int l1 = pairL.F;
            int l2 = pairL.S;
            int r1 = pairR.F;
            int r2 = pairR.S;
            vector<char> daddy = population[(randomRange(l1, r1))].moves;
            vector<char> mommy = population[(randomRange(l2, r2))].moves;
            
            if (l1 == 0 && l2 != 0)
            {
                for (int a = 0; a < chromossomeSize / 2; a++)
                {
                    if (a == 0)
                    {
                        daddy.push_back(directions[randomRange(0, 2)]);
                    }
                    else
                    {
                        daddy.push_back(coherentMoves[daddy.back()][randomRange(0, 2)]);
                    }
                }
                son = daddy;
            }
            else if (l2 == 0 && l1 != 0)
            {
                for (int a = 0; a < chromossomeSize / 2; a++)
                {
                    if (a == 0)
                    {
                        mommy.push_back(directions[randomRange(0, 2)]);
                    }
                    else
                    {
                        mommy.push_back(coherentMoves[mommy.back()][randomRange(0, 2)]);
                    }
                }
                son = mommy;
            }

            else
            {
                vector<char> first = mommy;
                vector<char> second = daddy;

                if (rand() % 1)
                {
                    swap(mommy, daddy);
                }

                int crossoverPoint = randomRange(1, chromossomeSize);

                for (int k = 0; k < crossoverPoint; k++)
                {
                    if (k < first.size())
                    {
                        son.push_back(first.at(k));
                    }
                }

                for (int k = crossoverPoint; k < chromossomeSize; k++)
                {
                    if (k < second.size())
                    {
                        if (k == crossoverPoint)
                        {
                            son.push_back(coherentMoves[son.back()][randomRange(0, 2)]);
                        }
                        else
                        {
                            son.push_back(second.at(k));
                        }
                    }
                }
            }
            string sonString(son.begin(), son.end());
            if (chromossomeSet.insert(sonString).S)
            {
                sonIsValid = true;
            }
        }
        individual newSon;
        newSon.moves = son;
        mutation(newSon);
        newPopulation.push_back(newSon);
    }

    return newPopulation;
}

/*
///////////////////////////////////////////////
Metodo main
*/

int main()
{
    randomSeed();
    printf(
"---------------------------------------------------------- \n\
#################### - MAZER SOLVER - #################### \n\
---------------------------------------------------------- \n"
            );
    cout << "Size of maze: ";
    cin >> mazeSize;
    mazeSize = 2 * mazeSize + 1;
    cout << endl;
    bool generate = true;
    do
    { 
        initVariables();
        mazeGenerator();
        drawMaze();
        cout << "Generate new maze?(y/n): ";
        string resposta;
        cin >> resposta;
        generate = (resposta == "y") ? true : false;
    } while(generate);
    
    cout << "Size of population: ";
    cin >> populationSize;
    cout << endl;
    
    initPopulation();
    int generation = 0; 
    int fittestFitness = 0;
    vector<char> fittest;

    int limitOfGenerations = 20000;
/*
    cout << "Limit of Generations: ";
    cin >> limitOfGenerations;
    cout << endl;
    */
    
    int start_s=clock();

    while(generation < limitOfGenerations){
        calculateFitness();
        fittest = getFittest();
        fittestFitness = population.front().fitness;
        
        ///if fittestFitness > fitnessConstant : finalists.append(fittest);
        ///print (len(finalists))

        population = crossover();
        
        generation += 1;

        //Sleep(500);        
        
        string fittestString(fittest.begin(), fittest.end());     
        cout << fittestString << endl;     
        //system ("CLS"); 
    }

    int stop_s=clock();
    cout << "time: " << (stop_s-start_s)/double(CLOCKS_PER_SEC)*1000 << endl;
   
    return 0;
}